posthog/models/app_metrics2/sql.py

```diff
@@ -0,0 +1,221 @@
+(
+    {BASE_APP_METRICS2_COLUMNS}
+    {KAFKA_COLUMNS_WITH_PARTITION}
+)
+ENGINE = {SHARDED_APP_METRICS2_TABLE_ENGINE()}
+PARTITION BY toYYYYMM(timestamp)
+ORDER BY (team_id, app_source, app_source_id, instance_id, {APP_METRICS2_TIMESTAMP_TRUNCATION}, error_type, error_uuid)
+"""
+)
+
+DISTRIBUTED_APP_METRICS2_TABLE_SQL = (
+    lambda: f"""
+CREATE TABLE IF NOT EXISTS app_metrics2 ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+(
+    {BASE_APP_METRICS2_COLUMNS}
+    {KAFKA_COLUMNS_WITH_PARTITION}
+)
+ENGINE={Distributed(data_table="sharded_app_metrics2", sharding_key="rand()")}
+"""
+)
+
+KAFKA_APP_METRICS2_TABLE_SQL = (
+    lambda: f"""
+CREATE TABLE IF NOT EXISTS kafka_app_metrics2 ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+(
+    team_id Int64,
+    timestamp DateTime64(6, 'UTC'),
+    app_source LowCardinality(String),
+    app_source_id String,
+    instance_id String,
+    successes Int64,
+    skipped Int64,
+    failures Int64,
+    error_uuid UUID,
+    error_type String,
+    error_details String CODEC(ZSTD(3))
+)
+ENGINE={kafka_engine(topic=KAFKA_APP_METRICS2)}
+"""
+)
+
+APP_METRICS2_MV_TABLE_SQL = (
+    lambda: f"""
+CREATE MATERIALIZED VIEW IF NOT EXISTS app_metrics2_mv ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'
+TO {settings.CLICKHOUSE_DATABASE}.sharded_app_metrics2
+AS SELECT
+team_id,
+timestamp,
+app_source,
+app_source_id,
+instance_id,
+successes,
+skipped,
+failures,
+error_uuid,
+error_type,
+error_details
+FROM {settings.CLICKHOUSE_DATABASE}.kafka_app_metrics2
+"""
+)
+
+TRUNCATE_APP_METRICS2_TABLE_SQL = f"TRUNCATE TABLE IF EXISTS sharded_app_metrics2"
+
+INSERT_APP_METRICS2_SQL = """
+INSERT INTO sharded_app_metrics2 (
+    team_id,
+    timestamp,
+    app_source,
+    app_source_id,
+    instance_id,
+    successes,
+    skipped,
+    failures,
+    error_uuid,
+    error_type,
+    error_details,
+    _timestamp,
+    _offset,
+    _partition
+)
+SELECT
+    %(team_id)s,
+    %(timestamp)s,
+    %(app_source)s,
+    %(app_source_id)s,
+    %(instance_id)s,
+    %(successes)s,
+    %(skipped)s,
+    %(failures)s,
+    %(error_uuid)s,
+    %(error_type)s,
+    %(error_details)s,
+    now(),
+    0,
+    0
+"""
+
+QUERY_APP_METRICS2_DELIVERY_RATE = """
+SELECT app_source, app_source_id, instance_id, if(total > 0, success/total, 1) as rate FROM (
+    SELECT app_source, app_source_id, instance_id, sum(successes) + sum(skipped) AS success, sum(successes) + sum(skipped) + sum(failures) AS total
+    FROM app_metrics2
+    WHERE team_id = %(team_id)s
+        AND timestamp > %(from_date)s
+    GROUP BY app_source, app_source_id, instance_id
+)
+"""
+
+# For composeWebhook apps we report successes and failures in two steps
+# 1. running the composeWebhook function
+# 2. rusty hook sending the webhook
+# Users don't care that there are two steps, we'll want to show them the
+# success count after step 2, but for failures we'll want to add them up
```

Comment:
https://posthog.slack.com/archives/C0374DA782U/p1718099036360589?thread_ts=1718012741.258499&cid=C0374DA782U

Ben will work on top of a fork of this branch and get back to us here, the question is:

in a hog application we can have multiple steps (e.g. imagine I do sth like
```
x = fetch(...)
if x.code = 200:
   fetch(...)
```
For reporting the data on the metrics page for this we'd want to count the last success (which is exit from hog or the last fetch), for failures we'd want to add all it up from all steps. Similar to composeWebhook now.

Should we add `step_number:int` there that will use a reserved `step_number = 0`  (or -1) in case it's the last step, and otherwise increasing value starting from 1, so we can do a query like this:
```
SELECT sumIf(successes, step_number = 0) AS successes, sum(failures) AS failures, sum(skipped) AS skipped FROM app_metrics2
WHERE ...
```
E.g. if the first fetch fails, then webhook delivery system reports it as `step_number=1`  (and it won't be passed forward to Hog, if it succeeds it's also reported with `step_number=1` and then passed forward to Hog), if the second fetch succeeds it would report it as `step_number=0` (step number would be passed into webhook delivery system)